src/scirpy/tl/_mutational_load.py

```diff
@@ -0,0 +1,203 @@
+from collections import defaultdict
+from collections.abc import Sequence
+from typing import Literal
+
+import numpy as np
+import pandas as pd
+
+from scirpy.get import airr as get_airr
+from scirpy.util import DataHandler
+
+
+def simple_hamming_distance(sequence: str, germline: str, frequency: bool, ignore_chars: list[str]):
+    if not sequence or not germline or len(sequence) != len(germline):
+        raise ValueError("Sequences might not be IMGT aligned")
+
+    distance = 0
+    num_chars = len(sequence)
+
+    for l1, l2 in zip(sequence, germline, strict=False):
+        if l1 in ignore_chars or l2 in ignore_chars:
+            num_chars -= 1
+
+        elif l1 != l2:
+            distance += 1
+
+    if num_chars == 0:
+        return np.nan  # can be used as a flag for filtering
+
+    elif frequency:
+        return distance / num_chars
+
+    else:
+        return distance
+
+
+@DataHandler.inject_param_docs()
+def mutational_load(
+    adata: DataHandler.TYPE,
+    *,
+    airr_mod="airr",
+    airr_key="airr",
+    chain_idx_key="chain_indices",
+    sequence_alignment: str = "sequence_alignment",
+    germline_alignment: str = "germline_alignment_d_mask",
+    chains: Literal["VJ_1", "VDJ_1", "VJ_2", "VDJ_2"] | Sequence[Literal["VJ_1", "VDJ_1", "VJ_2", "VDJ_2"]] = "VDJ_1",
+    region: Literal["IMGT_V(D)J", "IMGT_V_segment", "subregion"] = "IMGT_VDJ",
+    junction_col: str = "junction",
+    frequency: bool = True,
+    ignore_chars: list[str] = "None",
+    inplace: bool = True,
+) -> None | pd.DataFrame:
+    """\
+    Calculates observable mutation by comparing sequences with corresponding germline alignments and counting differences.
+    Needs germline alignment information, which can be obtained by using the interoperability to Dandelion (https://sc-dandelion.readthedocs.io/en/latest/notebooks/5_dandelion_diversity_and_mutation-10x_data.html)
+
+    Parameters
+    ----------
+    {adata}
+    {airr_mod}
+    {airr_key}
+    chain_idx_key
+        Key to select chain indices
+    sequence_alignment
+        Awkward array key to access sequence alignment information
+    germline_alignment
+        Awkward array key to access germline alignment information -> best practice mask D-gene segment (https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2)
```

Comment:
How do these information get there? Does dandelion / immcantation preprocessing populate these fields with "best practice" values? Does cellranger also provide them without any further processing?

Reply:
Yes, these columns are populated by such tools like Dandelion and Immcantation. In fact, it's the result of re-annotation with igBLAST or imgt/highv-quest that is run "under the hood" of these tools (see also https://immcantation.readthedocs.io/en/stable/getting_started/10x_tutorial.html#assign-v-d-and-j-genes-using-igblast). Both dandelion and Immcantation follow the AIRR Community Standard, meaning that both "sequence_alignment" and "germline_alignment" should always result in the IMGT-gapped sequence (see also https://immcantation.readthedocs.io/en/stable/datastandards.html)
As far as I am aware does cellranger not provide it in this format, which is the main reason why we have to re-annotate cellranger output in the first place 😢